Cython/Compiler/Optimize.py

```diff
@@ -382,6 +384,15 @@ def _transform_indexable_iteration(self, node, slice_node, reversed=False):
         body = Nodes.StatListNode(
             node.pos,
             stats = [target_assign]) # exclude node.body for now to not reanalyse it
+        if is_mutable:
+            # We need to be slightly careful here that we are actually modifying the loop
+            # bounds and not a temp copy of it. Setting is_temp=True on length_temp seems
+            # to ensure this.
+            # If this starts to fail then we could insert an "if out_of_bounds: break" instead
+            loop_length_reassign = Nodes.SingleAssignmentNode(node.pos,
+                                                        lhs = length_temp,
+                                                        rhs = copy.copy(length_call))
```

Comment:
I'm not a big fan of copying nodes. It works here, but only because special care was taken before, and future changes in other places can easily break these assumptions. I'd rather create a new node even here. Or push the  "mutable end boundary" down into the `ForFromStatNode`?

Reply:
I think pushing the mutable end boundary is probably a bigger change than I really want to do. I've created a new node.